adopt initial project governance (#19)
k8sgateway/community: GOVERNANCE.md

## Values

The K8sGateway project embraces the following values:

* Openness: Communication and decision-making happens in the open and is discoverable for future
  reference. As much as possible, all discussions and work take place in public
  forums and open repositories.
* Fairness: All stakeholders have the opportunity to provide feedback and submit
  contributions, which will be considered on their merits.
* Community over Product or Company: Sustaining and growing our community takes
  priority over shipping code or sponsors' organizational goals.  Each
  contributor participates in the project as an individual.
* Inclusivity: We innovate through different perspectives and skill sets, which
  can only be accomplished in a welcoming and respectful environment.
* Participation: Responsibilities within the project are earned through
  participation, and there is a clear path up the contributor ladder into leadership
  positions.

## Roles and membership

Our governance builds on top of definitions from other documents, review the following files for relevant information:

> * [ROLES.md](./ROLES.md) for the different roles with the project.

Author: roles is gone now on the tip of main. it's contrib-ladder now @linsun should I handle the rebase and update this or do you want to?

Reviewer: I got it : good eye catch!

Reviewer: updated

* [MAINTAINERS.md](./MAINTAINERS.md) for the project maintainers, including their organizational affiliation.
* [Code of Conduct](./CODE-OF-CONDUCT.md) for our code of conduct.

## Voting

While most business in K8sGateway is conducted by "[lazy consensus](https://community.apache.org/committers/lazyConsensus.html)",
periodically the Maintainers may need to vote on specific actions or changes. The following rules govern our voting process, unless otherwise stated for a specific purpose.

> * A vote begins when a PR that has the `/vote` comment.

Reviewer 2: ```suggestion
* A vote begins when someone adds a PR comment containing the `/vote` command.
```

> * A vote begins when a PR that has a comment with the label `up-for-vote` is submitted.

Reviewer 3: nit: `A vote begins with...`

Reviewer 3: I thought we decided on using a dedicated GH Issue to call a vote and to xref all necessary issues, PR, etc. that generated the call to vote?

Reviewer 2: per our discussion this week, we thought it was a good idea to call for votes via a comment within a PR, because it easily shows the time at which the vote was taken, and if further commits are pushed after that, it would invalidate the vote and a new vote could be called via a new comment on the PR

Reviewer: BTW, we have a vote bot integrated now, following CNCF TOC's best practice :). You can see a test issue here: https://github.com/k8sgateway/community/issues/39

Reviewer: Re [nit: A vote begins with...

> nit: `A vote begins with...`](nit: A vote begins with...), might be my english, i really think it is `when` here because it is a sentence after it.

> * A vote begins when a PR that has a comment with the label `up-for-vote` is submitted.

Reviewer 2: what does "a comment with the label `up-for-vote`" mean? maybe it meant to say that the PR (not the comment) contains the label?

there seems to also be a label called `do-not-merge/vote-needed` which seems similar to the other, do we need both?

> * A vote begins when a PR that has a comment with the label `up-for-vote` is submitted.

Reviewer 2: can we also be specific on how a call for vote is initiated via a comment? does the comment need to contain specific text, or just something to indicate that they are calling for a vote?

Reviewer: Updated, PTAL @jenshu @ilrudie

> * A simple majority decides the vote outcome.

Reviewer 4: Should we require that votes be received from more than one org to pass, in pursuit of the vendor-neutrality objective?

Author: For clarity, any vote or a vote in favor from multiple companies?

Consider a vote which ends with:
* company A: 4 up votes
* company B: 1 down vote

Is this a passing vote or a failed vote?

Reviewer: Updated to be eligible votes which a company can have a max of 6 eligible votes. This does mean it has to get more votes from more than one org to pass based on the eligible voters today.

Reviewer: To Ian's question, that is a passing vote in that case. But the issue with your example is the diversity of eligible voters which is why we try to recruit diversified maintainers as much as possible.

Author: I think my example is not so clear cut though. In this case we have 1 consenting company and 1 against. A simple majority is reached, plainly, however if we require 2 companies as suggested what is the outcome of this vote in that case? This is the scenario I'd like to consider. Is it "simple majority and at least 2 companies agree to it" or "simple majority if any 2 companies participate". We should also consider what happens if the vote goes to time with only a single company participating if we want to adopt verbiage around multiple companies...

We shouldn't IMO always assume all eligible voters will vote when considering how the processes will work so my hypothetical is a possible vote outcome for this project, as the maintainer council looks today.

> * A simple majority decides the vote outcome.

Reviewer 5: Simple majority of possible votes? 
The problem now is that if there exists a company (such as during this bootstrap phase Solo.io) that has more than the cap of 6 votes this gets a little weird.
I would argue that this should be Simple majority of possible votes and if not we minimally need to expand on this comment.

Ie if there are 16 total maintainers with 8 from org A then there are actually only 14 votes. In this case I would expect 8 valid votes to be able to pass instead of 9.

Author: Yeah, majority of "eligible votes" is maybe the way to put it?

Reviewer: Agreed, it needs to be "eligible votes".

> * A simple majority decides the vote outcome.

Reviewer: ```suggestion
* A simple majority of eligible votes decides the vote outcome.
```

> * A simple majority of eligible votes decides the vote outcome.

Reviewer 3: Consider defining "eligible".

Reviewer: It is currently defined in github as a group - see https://github.com/k8sgateway/community/blob/main/.gitvote.yml

> * A vote is cast when a project maintainer emotes thumbs-up or thumbs-down (👍 or 👎) on the GitHub "call to vote" issue.

Reviewer 6: Given that the emote can be changed or removed at any point without any historical view of previous votes, this seems like a risk we may want to address

I think it's safer to have the entire history; have we considered voting using any other format? For example, PR approvals for yes and changes requested for no. This also allows context on why a vote was cast the way it was (perhaps this omission is intentional)

Author: Does the risk of this kind of behavior only occur if there is an eligible voter (maintainer) looking to sabotage the project? I don't think this should really be an issue in practice if all maintainers are trustworthy. If all maintainers are not trustworthy we have serious, serious issues.

Reviewer: Hi @kdorosh thanks for weighing in on this.  We thought about this quite a bit and explored a few options:

1. Open an issue to call for vote and in that issue refer to the PR with a given commit.
2. Call vote directly in the PR, which btw is what the CNCF TOC repo uses to vote on projects as well, see an example: https://github.com/cncf/toc/pull/1473#issuecomment-2526194135. we could look into integrating a vote bot down the road but given we don't often, manual kicking off the vote should be okay.
- in this model, whenever there is a new commit (and you do see the history in the PR!), a new vote should be called with the latest change.
We decided to opt in for the 2nd method because it has everything in one PR and also used by the CNCF TOC repo which does vote a lot.

Reviewer: Updated the wording a bit to make it clear, ptal @kdorosh @ilrudie @tjons - thanks!

Reviewer 6: I think this works for now because we trust the maintainers. Dissenters can add reason for dissent in the issue as a new comment, so this can always be improved down the road. A vote bot may be preferable to depending on PR review semantics

I see PR review improving two scenarios:
1. Maintainer has a change of heart or is swayed to a new opinion
2. Maintainer has an obvious place to add dissenting reasons 

I care more about 2., but I think both can be improved in the future. I'm not going to block this PR on these improvements

> * A vote is cast when a project maintainer emotes thumbs-up or thumbs-down (👍 or 👎) on the GitHub "call to vote" issue.

Reviewer: ```suggestion
* A vote is cast when a project maintainer emotes thumbs-up or thumbs-down (👍 or 👎) on a specific "call to vote" comment of the GitHub PR, referring to the most recent commit when the comment is entered. If there is any further commit to the PR, a new vote must be called to refer to the new commit.
```

Author: Looks good. I think this was also implied, but making it explicit is a good idea

> * A vote is cast when a project maintainer emotes thumbs-up or thumbs-down (👍 or 👎) on the specific "vote" created by the git-vote bot of the GitHub PR, referring to the most recent commit when the vote is created. If there is any further commit to the PR, a new vote must be created.

Author: ```suggestion
* A vote is cast when a project maintainer emotes thumbs-up or thumbs-down (👍 or 👎) on the specific "vote" created by the git-vote bot on the GitHub PR, referring to the most recent commit when the vote is created. If there is any further commit to the PR, a new vote must be created.
```

> * A vote is cast when a project maintainer emotes thumbs-up or thumbs-down (👍 or 👎) on the specific "vote" created by the git-vote bot of the GitHub PR, referring to the most recent commit when the vote is created. If there is any further commit to the PR, a new vote must be created.

Reviewer 2: ```suggestion
* A vote is cast when a project maintainer emotes thumbs-up or thumbs-down (👍 or 👎) on the specific "vote" created by the git-vote bot of the GitHub PR, referring to the most recent commit at the time the vote is created. If there is any further commit to the PR, a new vote must be created.
```

Reviewer: Applied manually

* No single company should cast more than six votes. It is the responsibility of each company to determine which affiliated maintainers will vote.

> * Voting must be open for at least one week or until the required threshold is met.

Reviewer 3: We should set a time limit for a vote.

Reviewer: Yes it is set to a week, see https://github.com/k8sgateway/community/blob/main/.gitvote.yml

> * Voting must be open for at least one week or until the required threshold is met.

Reviewer 2: ```suggestion
* Voting must be open for at least one week or until the required threshold is met, whichever is sooner.
```

## Changes to governance

<!---
TODO: Enumerate what, precisely, constitutes a change in governance. I think it's clear editing this file, in any significant way, would count. Adjusting at least certain things wrt to the maintainer lifecycle also would likely constitute a governance change because this would directly impact who is, and is not, able to vote as well as who may be eligible or not in the future.
-->
All changes to our governance model require a vote by the maintainers. The proposed change will be submitted as a PR against the K8sGateway community repository. Voting will proceed according to the rules established for simple conflict resolution, with the sole exception being the threshold for acceptance.

> * Any change requires a 2/3 super majority to be accepted.

Reviewer 3: ^ also "eligible" votes?

Reviewer: I clarified eligible voters earlier, i think this is strongly implied.

> * Any change requires a 2/3 super majority to be accepted.

Reviewer 2: is this also max 6 votes per company?

Reviewer: Yes.

# K8sGateway governance

> 

Reviewer 3: Thoughts on including a charter? For example:

## Charter

The Steering Committee is the governing body of k8sgateway project, providing decision-making and oversight pertaining to project bylaws, sub-organizations, financial planning, etc. The Steering Committee is comprised of _____ and will be responsible for the following:

- Defining the project roadmap.
- Defining the project scope.

Author: I like the idea, I was really confused at first. I think the example given put me off since we don't want to have a steering committee but I think it's not intended to be actual verbiage adopted.

Author: I see some projects have an entire charter.md, would it make sense to break this out into an issue to create a charter.md and then link that back here?

Author: broken out into #23 to keep scope more limited

> K8sGateway is dedicated to building a feature-rich, fast and flexible ingress controller and next-generation API gateway. This governance document details how the project is run.

Reviewer 3: Ingress controller that k8sgw supports Ingress. IMHO the project should not support Ingress.

Author: I lifted this straight from somewhere in the project essentially verbatim. I do agree though it's odd verbiage given that we should not support the ingress resource.

  reference. As much as possible, all discussions and work take place in public
  forums and open repositories.
- Fairness: All stakeholders have the opportunity to provide feedback and submit
  contributions, which will be considered on their merits.
- Community over Product or Company: Sustaining and growing our community takes
  priority over shipping code or sponsors' organizational goals.  Each
  contributor participates in the project as an individual.
- Inclusivity: We innovate through different perspectives and skill sets, which
  can only be accomplished in a welcoming and respectful environment.
- Participation: Responsibilities within the project are earned through
  participation, and there is a clear path up the contributor ladder into leadership
  positions.

## Roles and membership

Our governance builds on top of definitions from other documents, review the following files for relevant information:

- [`ROLES.md`](./ROLES.md) for the different roles with the project.
- [`MAINTAINERS.md`](./MAINTAINERS.md) for the project maintainers, including their organizational affiliation.
- [Code of Conduct](./CODE-OF-CONDUCT.md) for our code of conduct.

> 

Reviewer 3: nit: Remove extra line

Author: done, also took a pass at fixing these lint type mistakes for the entire repo in #20

## Voting 

While most business in K8sGateway is conducted by "[lazy consensus](https://community.apache.org/committers/lazyConsensus.html)", 
periodically the Maintainers may need to vote on specific actions or changes. The following rules govern our voting process, unless otherwise stated for a specific purpose.

- A simple majority decides the vote outcome.

> - A vote is cast when a project maintainer emotes thumbs-up or thumbs-down (👍 or 👎) on the GitHub PR in dispute.

Reviewer 7: This specifies a GitHub PR. Does this apply to just PRs, or are GitHub issues also subject to votes?

Author: It's a good point. I think the discussion in community meetup today was to move to a "call for vote" issue which is a vanilla issue, has some context about why the vote is called and what is being voted on and has a label which makes it easy to search.

We had some trouble last time I would classify as "what are we voting on" since we voted on a PR, some folks reviewed it and left comments, which is great but not a vote. This lead to a sort of rolling vote problem where the PR was mutating during the vote which is sub-optimal. The "call for vote" issue linking out to the PR probably helps with that.

Author: Another plus to this, the issue is basically disposable. If the vote fails we can iterate an open PR or whatever we are voting on and then open a new "call for vote" instead of closing the PR and opening an entire new PR where the review context is lost just because we voted and the vote didn't pass.

Author: @danehans, would a "call to vote" generally occur in a central repository or in the repo being worked on? Like should "call to vote" be an org-level label or would those votes always occur in say, community, for convenience of finding all votes later on?

> - No single company should cast more than six votes. It is the responsibility of each company to determine which affiliated maintainers will vote.

Author: consider company block voting rather than making the company work out 6 individuals

Reviewer: 6 similar votes sound good, we have seen a few other successful projects in cncf doing the same.

> - Voting must be open for at least one week or until the required threshold is met.

Reviewer 7: When is voting considered "open"? From the time a particular issue or PR is opened? Is there a label that indicates which issues are being voted on or is there some other mechanism that I as a maintainer can know which issues are worth voting on? (should I just turn on all notifications for the repository)

Author: See comment above for context. Voting counts as open when the "call for vote" issue is submitted I think, this gives us a definitive date/time for start which is easy to track.

## Changes to governance

<!---
TODO: Enumerate what, precisely, constitutes a change in governance. I think it's clear editing this file, in any significant way, would count. Adjusting at least certain things wrt to the maintainer lifecycle also would likely constitute a governance change because this would directly impact who is, and is not, able to vote as well as who may be eligable or not in the future.
-->
All changes to our governance model require a vote by the maintainers. The proposed change will be submitted as a PR against the K8sGateway community repository. Voting will proceed according to the rules established for simple conflict resolution, with the sole exception being the threshold for acceptance.

- Any change requires a 2/3 super majority to be accepted.

> ## Vendor-neutrality

Reviewer 7: Is there a way to raise a concern about vendor-neutrality should it occur?

Author: I didn't make one. To avoid immediate escalation to the CNCF this is not a bad idea, although I suppose "bring it up with the CNCF" is a defacto option when none is given or when you find the given methods unsuitable.

Should we just have something like "Please raise concerns about neutrality by opening an issue or joining our community meeting." or is that too generic. Mailing list(not security so now we need a second one)? Slack?

Author: I did add a little blurb in the same vein as what I posted in the previous comment.
